gnomad_hail/utils/variant_qc.py

```diff
@@ -262,80 +274,476 @@ def quantiles_to_bin_boundaries(quantiles: List[int]) -> Dict:
+    singleton. Then for each grouping, min/max of `score` will be computed and any other desired aggregations.
+
+    Automatic aggregations that will be done are:
+        `min_score` - minimun of score annotation per group
+        `max_score` - maiximum of score annotation per group
+        `n` - count of variants per group
+        `n_ins` - count of insertion per group
+        `n_ins` - count of insertion per group
+        `n_del` - count of deletions per group
+        `n_ti` - count of transitions per group
+        `n_tv` - count of trnasversions per group
+        `n_1bp_indel` - count of one base pair indels per group
+        `n_mod3bp_indel` - count of indels with a length divisible by three per group
+        `n_singleton` - count of singletons per group
+        `fail_hard_filters` - count of variants per group with QD < 2 | FS > 60 | MQ < 30
+        `n_vqsr_pos_train` - count of variants that were a VQSR positive train site per group
+        `n_vqsr_neg_train` - count of variants that were a VQSR negative train site per group
+
+    Requires that `bin_ht` be annotated with an `info` struct that includes QD, FS, and MQ in order to add an
+    annotation for `fail_hard_filters`
+
+    :param bin_ht: Input Table with a `bin_id` annotation
+    :param agg_func: Function that returns a dict of any additional aggregations to perform
+    :param checkpoint_path: If provided an intermediate checkpoint table is created with all required annotations before shuffling.
+    :return: Table grouped by rank(s) and with counts of QC metrics
+    """
+    # Annotate binned table with the evaluation data
+    bin_ht = bin_ht.annotate(
+        indel_length=hl.abs(bin_ht.alleles[0].length() - bin_ht.alleles[1].length()),
+        fail_hard_filters=(bin_ht.info.QD < 2) | (bin_ht.info.FS > 60) | (bin_ht.info.MQ < 30)
+    )
+
+    # Explode the rank table by bin_id
+    bin_ht = bin_ht.annotate(
+        quantile_bins=hl.array([
+            hl.Struct(
+                bin_id=bin_name,
+                bin=bin_ht[bin_name]
+            )
+            for bin_name in bin_ht.bin_stats
+        ])
+    )
+    bin_ht = bin_ht.explode(bin_ht.quantile_bins)
+    bin_ht = bin_ht.transmute(
+        bin_id=bin_ht.quantile_bins.bin_id,
+        bin=bin_ht.quantile_bins.bin
+    )
+    bin_ht = bin_ht.filter(hl.is_defined(bin_ht.bin))
+
+    if checkpoint_path is not None:
+        bin_ht.checkpoint(checkpoint_path, overwrite=True)
+    else:
+        bin_ht = bin_ht.persist()
+
+    # Group by bin_id, bin and additional stratification desired and compute QC metrics per bin
+    return (
+        bin_ht.group_by(
+            bin_id=bin_ht.bin_id,
+            contig=bin_ht.locus.contig,
+            snv=hl.is_snp(bin_ht.alleles[0], bin_ht.alleles[1]),
+            bi_allelic=~bin_ht.was_split,
+            singleton=bin_ht.singleton,
+            release_adj=bin_ht.ac > 0,
+            bin=bin_ht.bin
+        )._set_buffer_size(20000)
```

Comment:
I know we discussed this yesterday and it sounded great, but I am struggling to implement this since when you return a GroupedTable it doesn't link back to the annotations that were on the table prior to grouping so I get an error. should we just put it all in one function like before and then name it `default_compute_grouped_binned_ht` or `default_compute_aggregate_binned_data`?

Reply:
Let's take this to Slack -- might be easier